doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,331 @@
+# Efficient Window Functions
+
+By “window functions”, this document means the `OVER` clause, e.g.,
+`SELECT row_number() OVER (PARTITION BY col1 ORDER BY col2) FROM relation;`
+
+**(Note that [temporal windows](https://materialize.com/docs/sql/patterns/temporal-filters/) are entirely different from what we discuss in this doc. Our support for those is already quite fine. There is no need to use SQL’s `OVER` clause for those.)**
+
+# Overview
+
+[Many users want to use window functions](https://www.notion.so/Window-Functions-Use-Cases-6ad1846a7da942dc8fa28997d9c220dd), but our current window function support is very inefficient: We recompute results for an entire window partition for any small change in the partition. So the only situations when our current support works is if the window partitions are either very small, or they rarely change.
+
+## Goals
```

Comment:
This section seems to mix goals with implementation, for example that some things are easy and others are not, or that `prefix_sum` will handle things for us. Ideally, this section lays out what we'll have after applying the change to Materialize that we didn't have before. The actual mapping should be part of the design section I believe.

Reply:
I mostly separated them, but I think a few short hints to which stuff will be supported by prefix sum vs. translating away window functions is still fine to have here.